do connection setup here

marten-seemann opened the pull request:
This package now sets up new connections. A go-libp2p-conn.Conn is an encrypted, stream-multiplexed connection. If the underlying tpt.Conn is a single-stream connection, it is first encrypted (using secio) and then a stream multiplexer is used to provide multistream support. If the tpt.Conn supports transport-level stream multiplexing, the secio handshake is performed on a new stream.

This PR depends on https://github.com/libp2p/go-libp2p-transport/pull/20.
There are still a couple of TODOs here, especially regarding the timeouts for dialing connections as well as the connection protector.

marten-seemann committed:
more consistent import names

marten-seemann committed:
setup new connections

This package now sets up new connections. A go-libp2p-conn.Conn is an
encrypted, stream-multiplexed connection. If the underlying tpt.Conn is
a single-stream connection, it is first encrypted (using secio) and then
a stream multiplexer is used to provide multistream support.

whyrusleeping commented on dial.go:
```diff
@@ -93,61 +98,71 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
-			}
-			maconn = pconn
-		}
+		// TODO: reenable the protector
```
@Kubuxu how would the protector work with something like QUIC?

marten-seemann replied:
Solved by https://github.com/libp2p/go-libp2p-pnet/pull/8.

whyrusleeping commented on dial.go:
```diff
@@ -93,61 +98,71 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 
-		maconn.SetReadDeadline(time.Now().Add(NegotiateReadTimeout))
+		var stream smux.Stream
+		if ssc, ok := tptConn.(tpt.SingleStreamConn); ok {
```
Use a type switch here:
```
switch con := tptConn.(type) {
case tpt.SingleStreamConn:
case tpt.MultiStreamConn:
default:
}
```

marten-seemann replied:
Done.

Kubuxu commented on dial.go:
```diff
@@ -2,24 +2,26 @@ package conn
 	"strings"
 	"time"
 
+	smux "github.com/jbenet/go-stream-muxer"
```
`s/jbente/libp2p/`

whyrusleeping commented on conn.go:
```diff
@@ -20,34 +23,103 @@ var log = logging.Logger("conn")
+		}
+		defer stream.Close()
+
+		secSession, err = setupSecureSession(ctx, local, privKey, stream)
```
Hrm... i'm not sure this is the right way to do this. This is encrypting on top of of the stream muxer, instead of underneath it.

I think that instead of doing it this way, we can just not support private networks and QUIC at the same time. Just error out instead i'm thinking.

cc @Kubuxu @lgierth

whyrusleeping replied:
oh wait, this isnt the private networks protector layer. i was mistaken.

marten-seemann replied:
The problem here was that I wasn't using the `ReadWriter` provided by the secio package. I fixed that.

whyrusleeping commented on conn.go:
```diff
@@ -20,34 +23,103 @@ var log = logging.Logger("conn")
+
+	var streamConn smux.Conn
+	var secSession secio.Session
+	if ssc, ok := tptConn.(tpt.SingleStreamConn); ok {
```
you can do a type switch here:
```go
switch sc := tptConn.(type) {
case tpt.SingleStreamConn:
  // sc is a SingleStreamConn in this case block
case tpt.MultiStreamConn:
  // sc is a multistreamconn in this case block
default:
  return errWrongType
}
```

marten-seemann replied:
Done.

Kubuxu commented on listen.go:
```diff
@@ -149,78 +153,91 @@ func (l *listener) handleIncoming() {
+			} else if msc, ok := conn.(tpt.MultiStreamConn); ok {
+				stream, err = msc.AcceptStream()
+				if err != nil {
+					panic(err)
```
We can't panic here?

Kubuxu commented:
The mixing of the secio with the multiplexed connections looks good, what we also have to do is confirm that person on the other side of the secio is using the exactly the same QUIC connection.

This is to prevent someone proxying the secio over different QUIC. This is a problem as confirming identity over secio over QUIC doesn't prove the identity of the the other side of QUIC.

whyrusleeping commented on conn.go:
```diff
@@ -20,34 +23,110 @@ var log = logging.Logger("conn")
+			return nil, err
+		}
+	case tpt.MultiStreamConn:
+		// 1. secure the connection on the first best stream
```
I think we want to add a big TODO and WARNING here, we're only doing this this way for now, for quic, until we have the ability to simply rely on QUICs crypto to validate the peers. Its also insecure (can be man in the middled) so we should note that QUIC is experimental for now.

whyrusleeping commented on dial.go:
```diff
@@ -93,61 +98,71 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
+		}
 
-		err = msmux.SelectProtoOrFail(cryptoProtoChoice, maconn)
+		stream.SetReadDeadline(time.Now().Add(NegotiateReadTimeout))
```
should check the error here

whyrusleeping replied:
(oh, i guess this wasnt checked before, ignore this comment)

whyrusleeping commented on dial.go:
```diff
@@ -93,61 +98,71 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 
-		maconn.SetReadDeadline(time.Time{})
+		// clear deadline
+		stream.SetReadDeadline(time.Time{})
```
error check again

whyrusleeping commented on dial.go:
```diff
@@ -93,61 +98,71 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 		if d.Protector != nil {
-			pconn, err := d.Protector.Protect(maconn)
+			var pconn tpt.Conn
+			pconn, err = d.Protector.Protect(tptConn)
```
So in this case, if using a 'protector', we end up just stream multiplexing over a single QUIC conn, am i reading that correctly?

If that is the case, then we might just want to (for now) disable the use of QUIC and private networks at the same time.

marten-seemann replied:
The protector should work fine.
`tptConn` is the general `tpt.Conn`. `pnet.Protect` does another type assertion in https://github.com/marten-seemann/go-libp2p-pnet/blob/11a8850893ddc8d571abe57b5082d0110b77d412/protector.go#L33-L42. For `MultiStreamConn`s, it wraps `OpenStream` and `AcceptStream` and uses the PSK with a separate nonce for every stream (see https://github.com/marten-seemann/go-libp2p-pnet/blob/11a8850893ddc8d571abe57b5082d0110b77d412/psk_multistreamconn.go#L22-L36).

whyrusleeping commented:
So to move forward here, I think we want to disable private networks and QUIC being used together, as well as adding a few more comments about the security implications of the secio peerID verification we're doing.

Also, to make future integration easier, I think we should use the servers peer private key as the QUIC cert, even though this doesnt buy us any security right now, it will help make the interfaces clearer for when we are able to rely on tls to do the client<->server verification (aka, when we can trust quics crypto to do the peer handshake for us).

marten-seemann committed:
remove conn.Serve

This method was recently removed from the interface.

marten-seemann commented:
I think there is an easy (although not very aesthetic) solution for the MITM vulnerability for QUIC connections (I justed pushed that one). The QUIC vulnerability arises because secio only secures the stream the handshake is performed on. All other streams are unencrypted. A MITM would just forward the secio stream and then intercept the data streams.
What we can do here is use the `Protector` to encrypt the data streams. The PSK needed to do that could is generated by server (it doesn't matter who does that) and sent as the first thing on the secio-secured stream.

I'm really looking forward to get rid of all those hacks. I recently started with the TLS integration in quic-go, but it will probably take me some time to get that production-ready.

@whyrusleeping: I don't understand your command about the certs. I guess you mean something that we establish a CA to sign the server's private key and use that cert as the cert for the QUIC connection?

marten-seemann committed:
reenable the tests, and add some more

marten-seemann committed:
remove dead code

Those functions aren't used anywhere, so this is not expected to break
anything. They also don't belong into this package anyway.

marten-seemann committed:
introduce a timeout for dialing and accepting connections

This removes the timeout for the multistream selection, and implements a
timeout covering the whole dialing and accepting process. Timeouts can
occur at multiple places during connection establishments.

marten-seemann committed:
use the renamed transport interfaces

marten-seemann commented:
@Stebalien: Could you review this please? This could be merged now that the new transport interfaces are merged.

Stebalien commented on .travis.yml:
```diff
@@ -11,6 +11,8 @@ install: true
 before_install:
   - make deps
+  - go get github.com/onsi/ginkgo/ginkgo
+  - go get github.com/onsi/gomega
```
These should get pulled in by the `make deps` target (so `make deps; gx test ./...` works on a clean install). That is,


```make
ginkgo:
	go get github.com/onsi/ginkgo/ginkgo
	go get github.com/onsi/gomega

deps: gx covertools ginkgo
...
```

Stebalien commented on secure_conn.go:
```diff
@@ -1,130 +1,20 @@
-// LocalMultiaddr is the Multiaddr on this side
-func (c *secureConn) LocalMultiaddr() ma.Multiaddr {
-	return c.insecure.LocalMultiaddr()
+	tpt.DuplexConn               // the wrapped conn
```
Let's not expose this if at all possible.

marten-seemann replied:
What do you mean?

Stebalien replied:
This exports DuplexConn's exported methods on `secureConn`. I'm worried that users may try casting `iconn.Conn` to `DuplexConn`, calling methods on the underlying insecure connection.

marten-seemann replied:
Done.

Stebalien commented on conn.go:
```diff
@@ -2,52 +2,127 @@ package conn
+	}
+	// force the handshake right now
+	// TODO: find a better solution for this
+	b := []byte("handshake")
```
Why do we do this? I assume this has to do with verifying multiplexed connections. If so, this should be done in the MultiplexConn branch of newSingleConn (i.e., we shouldn't do this for non-multiplex connections).

marten-seemann replied:
I had to force the handshake to get QUIC working. I'm not entirely sure why it didn't work otherwise (I never understood why the secio handshake is delayed), I'd have to debug this again.

Since QUIC support will not be merged before TLS 1.3, we might not need this at all. I will just remove this code for now.

marten-seemann replied:
It's a bit more complicated. The code won't work for duplex conns if we remove this here, because the context passed to secio might be cancelled before the handshake completes.

I'm wondering why we need this delayed handshake at all, it doesn't give us any benefit here, but makes it a lot harder to reason about timeouts (if the handshake is delayed, we can't have a connection establishment timeout).

Stebalien replied:
I'm not sure why we do this lazily but this definitely isn't the solution (it would, at the very least, require a protocol version bump to avoid breaking existing implementations).

I believe you can currently force a secio handshake by calling `secSession.ReadWriter()` (ideally, you'd be able to write an empty byte string to force the issue but we tend not to handle those very well in libp2p...).

whyrusleeping replied:
Hrm... there *was* a reason at some point. But i cant recall, and i'm not sure if its still valid. @Kubuxu do you remember?

marten-seemann replied:
@Stebalien: `secSession.ReadWriter()` forces the handshake, but it doesn't return the error, which we'd definitely want for logging purposes. As a workaround, we could add an error method to the secio `closedRW` and retrieve the error message by type asserting. We wouldn't have to send anything on the wire then, so we could avoid the protocol version bump.

I'd be happy to refactor secio such that the handshake is performed immediately, if this feature isn't needed anymore. I think this would be the cleanest solution here (a handshake takes a couple of RTTs, so latency-wise we're better off doing it asap).

marten-seemann replied:
I just created a PR in the secio repo to remove the lazy handshake: https://github.com/libp2p/go-libp2p-secio/pull/16.

whyrusleeping replied:
We should be able to remove this now. This PR should work with or without the lazy handshake, and with or without this hacky forced handshake thing.

whyrusleeping replied:
Also, for reference, you can just call `.RemotePeer()` to force the handshake.

marten-seemann replied:
This shouldn't be needed anymore at all. I just forgot to remove this. PR is updated now.

Stebalien commented on conn.go:
```diff
@@ -2,52 +2,127 @@ package conn
+			return nil, err
+		}
+	case tpt.MultiplexConn:
+		// 1. secure the connection on the first best stream
```
Is it possible to get a hold of the server's public QUIC key? After thinking about this, having the client send the server it's own public key over the secio connection should allow us to verify the outer QUIC connection.

Alternatively, if the server uses their secio key *as* their QUIC key (converts it), simply establishing a secio connection and sending a ping back and forth should be sufficient.

Stebalien replied:
We can also do as you suggest and treat each stream as a separate connection and protect each one.

Regardless, whatever we do, this has implications on the transport interfaces. IMO, we should require that *all* multiplexed connections provide their own security (which means we may need to change that interface...).

Stebalien replied:
Actually, if we *do* assume that all multiplexed transports are secure, we should probably push the security stuff into the QUIC smux library.

marten-seemann replied:
I like your last idea. I think it's a reasonable assumption that all multiplexed transports have their own security mechanism (QUIC is the first multiplexed transport we have, and all future ones will only get better than that).

I started working on the TLS 1.3 integration into QUIC last week, and I have a very basic version working now (still far from production use though). Once I'm done with this, this will allow us to get rid of secio (and probably also the protector, if we want to).

We still have to decide what we do here. I'm not sure what a good solution would look like. Maybe we should just panic here.

Stebalien replied:
For now, panicing sounds reasonable; we don't want users using this at this point. We'll need to fix the MultiplexConn interface regardless of what we do (to add security related methods) so it isn't really usable as-is anyways.

---

After thinking about all this, I think there may be a better way to structure this (but we can do this later unless you think this will be quick and easy). Basically, we can make *all* transports return multiplexed, secured connections and then provide convenience methods to provide security/multiplexing for protocols that don't have built-in support. For example, the TCP transport would be defined as:

```go
type tcpDialer struct {/* ... */}

func (t *tcpDialer) Dial(raddr ma.Multiaddr) {
    c, err := tcp.Dial(...)
    if err != nil {
        return err
    }
    // wraps the connection c using a stored secio generator.
    sc, err := t.secioSession.Wrap(c)
    if err != nil {
        return err
    }
    return smux.Multiplex(sc) // negotiates the multiplexer.
}
```

IMO, this better fits go's composition over inheritance pattern.

Stebalien commented on conn.go:
```diff
@@ -58,7 +133,10 @@ func (c *singleConn) Close() error {
+	if err := c.streamConn.Close(); err != nil {
+		return err
+	}
+	return c.tptConn.Close()
```
Does it hurt to close this regardless? I'd rather not leave connections around because of a problem with the muxer.

```go
serr := c.streamConn.Close()
terr := c.tptConn.Close()

if terr != nil {
    // If closing the stream fails, it's probably because there was a problem with the transport. Ideally, we'd be able to combine these errors but that may make it hard for higher levels to check for known, standard errors (e.g., `io.EOF`).
    return terr;
} else {
    return serr;
}
```

marten-seemann replied:
Done. 

I think there's an ambiguity in the stream-muxer interface, it's not clear what `Close` is supposed to do:
1. close the stream muxer and the underlying connection
2. only close the stream muxer, not the underlying connection

The change we just did assumes 1.

marten-seemann replied:
We should decide for one option in https://github.com/libp2p/go-stream-muxer/pull/19 and adjust this here accordingly.

Stebalien replied:
Ah. Good point. I believe closing the stream muxer is generally sufficient.

Stebalien commented on dial.go:
```diff
@@ -3,23 +3,36 @@ package conn
+var DialTimeout = 60 * time.Second
+
+// dialTimeoutErr occurs when the DialTimeout is exceeded.
+type dialTimeoutErr struct{}
```
Any reason we don't export this?

marten-seemann replied:
Doing a type assertion for `net.Error` should give you all the information necessary, and I think it's the more idiomatic way anyway.

Stebalien replied:
In go, you're probably right...

Stebalien reviewed (commented):
Partial review so we can get the discussion started.

whyrusleeping commented on listen.go:
```diff
@@ -149,78 +138,112 @@ func (l *listener) handleIncoming() {
+					defer stream.Close()
+				}
+
+				if _, _, err := l.mux.Negotiate(stream); err != nil {
```
Definitely unrelated to this PR, but could we check that the protocol chosen here is secio, and error out otherwise? It would be helpful if nodes on older code don't behave weirdly when we start supporting tls1.3

whyrusleeping commented:
@marten-seemann i'm fine moving forward with parts of this PR, if we can do the MultiplexConn stuff in a followup PR, i think we can get the rest in much more quickly.

marten-seemann committed:
remove the MultiplexConn type assertions

marten-seemann commented:
I removed the MultiplexConn type assertions (and everything related with that), so we could get the rest merged (this will also require merging another change to go-libp2p-transport).

Stebalien commented on dial.go:
```diff
@@ -69,112 +84,91 @@ func (d *Dialer) String() string {
 			cryptoProtoChoice = NoEncryptionTag
 		}
 
-		maconn.SetReadDeadline(time.Now().Add(NegotiateReadTimeout))
```
Shouldn't we still be setting this (or, more generally, a timeout on the entire connection)? Otherwise, we could wait here *forever*. Yes, we'll move on elsewhere but we'll never free these resources.

Maybe change this to the dial timeout (and set both the read and write deadlines)/raw

marten-seemann replied:
This should be taken care of by the context. When the context times out, the `rawConn` is closed, causing the `Read` to return immediately.

Stebalien replied:
Ah, good point.

Stebalien commented on conn.go:
```diff
@@ -2,50 +2,89 @@ package conn
+		c = &secureConn{
+			insecure: tptConn,
+			secure:   secSession,
+		}
```
If `remote` is not `""`, we should check if `remote == secSession.RemoteID()` here. That way, we do this *before* doing anything else. Currently, we check in swarm but, IMO, that's too late. Note: this is an improvement, not a bug.

marten-seemann replied:
Can we do that in a separate PR?

Stebalien replied:
Fair enough.

Stebalien commented on listen.go:
```diff
@@ -74,41 +80,24 @@ func (l *listener) SetAddrFilters(fs *filter.Filters) {
-		if l.privk == nil || !iconn.EncryptConnections {
-			log.Warning("listener %s listening INSECURELY!", l)
-			return c, nil
+	for c := range l.incoming {
```
Don't need a loop. We had one for the `IsTemporary` error check but that now happens lower down.

This can be simplified to:

```go
c := <- l.incoming
return c.conn, c.err
```

marten-seemann replied:
Fixed.

Stebalien commented on secure_conn.go:
```diff
@@ -1,128 +1,60 @@
 	ma "github.com/multiformats/go-multiaddr"
 )
 
-// secureConn wraps another Conn object with an encrypted channel.
```
Comment to be reverted.

marten-seemann replied:
Done.

Stebalien commented on listen.go:
```diff
@@ -149,78 +138,98 @@ func (l *listener) handleIncoming() {
-			if l.protec != nil {
-				pc, err := l.protec.Protect(maconn)
+
+			ctx, cancel := context.WithTimeout(l.ctx, ConnAcceptTimeout)
```
We should consider setting a `ConnAcceptTimeout` deadline on the connection itself to avoid leaking go routines.

Stebalien replied:
As with dialing, the connection will be closed when this expires so ignore this.

Stebalien reviewed (changes requested):
Nice. I have a few comments that I can fix myself depending on your timing but this is mostly good to go.

Stebalien commented:
Also, I'll go ahead and update the secio dep version.

marten-seemann committed:
simplify accepting new connections

Stebalien commented on listen.go:
```diff
@@ -74,41 +79,19 @@ func (l *listener) SetAddrFilters(fs *filter.Filters) {
 	}
-	return nil, fmt.Errorf("listener is closed")
+
+	c := <-l.incoming
```
Actually, I was wrong. We'll also need to check if the channel is closed (`c, ok := <-l.incoming`...).

marten-seemann replied:
Right. Fixed.

Stebalien commented:
So, apparently this still does multiplexing setup and needs https://github.com/libp2p/go-libp2p-interface-conn/pull/1 (which we can't merge as it depends on https://github.com/libp2p/go-libp2p-transport/pull/20 which we reverted).

Stebalien commented on listen.go:
```diff
@@ -149,78 +135,98 @@ func (l *listener) handleIncoming() {
+				}
+
+				// If we have a wrapper func, wrap this conn
+				if l.wrapper != nil {
```
This swaps the order of wrapping and protecting. Was this a bug?

Stebalien replied:
Yes, I believe it was. We wrap/protect in *this* order when dialing.

Stebalien commented on listen.go:
```diff
@@ -149,78 +135,98 @@ func (l *listener) handleIncoming() {
+			}()
+
+			select {
+			case <-ctx.Done():
```
There's a race here. We can return the connection and then close it.

marten-seemann replied:
I pushed a fix for that.

Stebalien commented on listen.go:
```diff
@@ -149,78 +135,98 @@ func (l *listener) handleIncoming() {
+				}
+
+				if _, _, err := l.mux.Negotiate(conn); err != nil {
+					log.Warning("incoming conn: negotiation of crypto protocol failed: ", err)
```
Wrong error.

Stebalien replied:
Actually, what exactly is this doing? You didn't change this...

Stebalien replied:
Ah. This is negotiating security (we overload mux way too much). Nevermind.

marten-seemann committed:
fix race condition when accepting connections

marten-seemann commented:
@Stebalien: Is there anything else I need to fix here?

Stebalien commented:
The only thing blocking this is https://github.com/libp2p/go-libp2p-conn/pull/9#issuecomment-339879541. I've been working on a new PR with the secio/async stuff so we can move that forward (I can hand it off to you if you want, I'm hitting some odd test failures around timeouts that you've probably already figured out). I'm also working on the transport interfaces problems but it doesn't look like it's going to be quite as simple as I thought... (which is why I'd like to just get the async accept stuff merged first).

Stebalien commented:
That is, #18
